src/loader/mock.ts

```diff
@@ -0,0 +1,82 @@
+/// <reference path="../../node_modules/@dojo/loader/dojo-loader.d.ts" />
+
+import { createHandle } from '@dojo/core/lang';
+import { Handle } from '@dojo/interfaces/core';
+import { forOf } from '@dojo/shim/iterator';
+import Map from '@dojo/shim/Map';
+
+declare const define: DojoLoader.Define;
+
+const moduleMap = new Map<string, () => void>();
+
+let disableHandle: Handle | undefined;
+
+/* TODO: remove when https://github.com/dojo/loader/issues/126 is resolved */
+declare global {
+	interface NodeRequire {
+		undef(moduleId: string, recursive?: boolean): void;
+	}
+}
+
+/**
+ * Register a module to be mocked when mocking is enabled
+ * @param mid The absolute module ID to mock
+ * @param mock The mock definition of the module
+ */
+export function register(mid: string, mock: any, req: NodeRequire = require): Handle {
+	if (disableHandle) {
+		throw new Error('Cannot register modules while mock is enabled.');
+	}
+	const absoluteMid = req.toAbsMid(mid);
+	moduleMap.set(absoluteMid, () => define(absoluteMid, [], () => mock));
+	return createHandle(() => {
+		if (disableHandle) {
+			req.undef(absoluteMid);
+			require.cache({
+				[mid]: undefined
+			});
+
+			/* TODO: remove when https://github.com/dojo/loader/issues/124 resolved */
```

Comment:
PR is already open for loader issue 124, so maybe review that PR and land and then update this PR?